### ISSUE TITLE
styled-components v5.2 broke proper encapsulation

### ISSUE THREAD

**Issue by lifeiscontent** [styled-components/styled-components#3334]
## Environment

`npx envinfo --system --binaries --npmPackages styled-components,babel-plugin-styled-components --markdown --clipboard`

## System:
 - OS: macOS 11.0.1
 - CPU: (12) x64 Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
 - Memory: 311.99 MB / 16.00 GB
 - Shell: 5.8 - /bin/zsh
## Binaries:
 - Node: 12.19.0 - ~/n/bin/node
 - Yarn: 1.18.0 - ~/.yarn/bin/yarn
 - npm: 6.14.8 - ~/n/bin/npm
 - Watchman: 4.9.0 - /usr/local/bin/watchman
## npmPackages:
 - styled-components: 5.2.1 => 5.2.1

## Reproduction

https://codesandbox.io/s/sad-hoover-0psm9?file=/src/index.js

## Steps to reproduce

look at the reproduction. this test case is working as expected in V4 and V5.1

## Expected Behavior

the component should render extended styles based off state but doesn't (look at comments in reproduction for specifics)

## Actual Behavior

the component renders whatever styles win in specificity (due to cascade)

**Comment by rikutiira** [comment]
This broke our styles in pretty major ways and we had to downgrade back to 5.2.0 from 5.2.1.

**Comment by JustFly1984** [comment]
I can confirm the bug. 
In the project I use for my client we are using framer-motion, which itself creates it's own react context, and passing it down by the component tree. It works fine with 5.2.0, but after update to 5.2.1, animations execution invocation is broken due to framer-motion context being not forwarded by the styled-components anymore.

It works ok while all the code is in the same huge 1000 lc component, but then I started to extract JSX with animations to separate components (for perf optimizations) I found this bug. 

Thanks to @rikutiira for temp solution.

**Comment by lifeiscontent** [comment]
ping @probablyup

**Comment by cyco130** [comment]
I believe I ran into the same issue. I have a more minimal reproduction here: https://codesandbox.io/s/styled-components-52-regression-jydzc

**Comment by probablyup** [comment]
This is my proposal for fixing the issue https://github.com/styled-components/styled-components/issues/3244#issuecomment-687677994

**Comment by probablyup** [comment]
If you do `&&` instead of `&` I think it should fix it in the meantime. https://codesandbox.io/s/distracted-wozniak-mxesf?file=/src/index.js

**Comment by anark** [comment]
I don't think using `&&` is a proper solution for this.

A simple demonstrations can be seen with the following
```
import React from 'react';
import styled, { css } from 'styled-components';

const ListItem = styled.div`
  ${({ selected }) =>
    selected &&
    css`
      .dark & {
        background-color: red;
      }
    `}
`;

export const App = () => (
  <div className="dark">
    <ListItem selected>Should be red</ListItem>
    <ListItem>This shouldn't be red</ListItem>
  </div>
);
```

In this example, if there is a single.selected item then all items will have the red background color.  This is not the expected behaviour and this did not happen prior to 5.2.

**Comment by arimah** [comment]
I think there is a huge issue with the way `&` is implemented in 5.2. It's not that `&&` is a special sequence that expands to "all instances plus current instance"; it's that the second `&`, _no matter where it appears_, means "current instance". To build on @cyco130's example, you could tweak the `Text` components like so:

```js
const Text = styled.span`
  ${Button}:hover &,
  ${Button}:focus & {
    color: ${(props) => (props.hoverable ? "red" : "inherit")};
  }
`;
```

... and now the `:focus` styles (as expected) get the props-dependent styling, while `:hover` does not. This of course means it's impossible to use `&` to mean "all instances" more than once, unless you use a bunch of `css` blocks.

As I see it, there are three reasonable ways to remedy this situation:

1. Normalise `&` so that it _only_ refers to all instances, and special-case `&&` (with _no intervening tokens_, not even white space) to mean what @probablyup keeps suggesting we use it for, or add another selector for "current instance".
2. Revert `&` to its previous behaviour and add a new selector to mean "all instances". Using `&&` for this seems problematic, since it would presumably have higher specificity, whereas it's usually more desirable for instance styles to override.
3. Revert `&` to its previous behaviour and have no special syntax for "all instances".

Perhaps one could also add a dev warning if multiple instances try to apply conflicting styles to a block with an "all instances" selector.

For (3), there is an ugly way to get an "all instances" selector, as follows (I'm sure linters will love this):

```js
const Child = styled.span`
  ${Parent} ${() => Child} {
    ...
  }
`;
```

Please share your thoughts. I think Styled Components is far too good of a library to let (what I think is) a design misstep hold people back from building great websites. :)

**Comment by kitten** [comment]
Duplicate of #3265
